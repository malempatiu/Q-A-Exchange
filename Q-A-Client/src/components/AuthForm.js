import React, { Component } from 'react';
import { connect } from 'react-redux';
import { userAuthentication, addError } from '../actions/actions';
import { Redirect, Link } from 'react-router-dom';

export class AuthForm extends Component {
    state = {
        email: '',
        username: '',
        password: ''
    };
    handleEmail = (e) => {
        this.setState({ email: e.target.value });
    };
    handlePassword = (e) => {
        this.setState({ password: e.target.value });
    };
    handleUsername = (e) => {
        this.setState({ username: e.target.value });
    };
    handleFormSubmit = (e) => {
        e.preventDefault();
        const { signUp, userAuthentication } = this.props;
        const authType = signUp ? 'signup' : 'signin';
        if(this.state.password.length>=6){
        userAuthentication(authType, this.state);
        //document.getElementById('auth-form').reset();
        }else{
            this.props.addError("Check password length");
            //document.getElementById('auth-form').reset();
        };
    };
    render() {
        const { isAuthenticated } = this.props;
        if (isAuthenticated) {
            return <Redirect to="/" />
        }
        const { signUp, btnText, heading, error } = this.props;
        return (
            <div>
                <div className="container authform-cont">
                    <h1 className="text-center">{heading}</h1>
                    {error && <div className="alert alert-danger">{error}</div>}
                    <div className="row">
                        <div className="col-lg-12 col-md-12 col-sm-12">
                            <form id="auth-form" onSubmit={this.handleFormSubmit}>
                                <div className="form-group">
                                    <label className="font-weight-bold" htmlFor="email">Email:</label>
                                    <input type="text" className="form-control" id="email" onChange={this.handleEmail} />
                                    <small id="emailHelp" className="form-text text-muted">abc@abc.com</small>
                                </div>
                                <div className="form-group">
                                    <label className="font-weight-bold" htmlFor="password">Password:</label>
                                    <input type="password" className="form-control" id="password" onChange={this.handlePassword} />
                                    <small id="emailHelp" className="form-text text-muted">min.length 6</small>
                                </div>
                                {signUp && (
                                    <div className="form-group">
                                        <label className="font-weight-bold" htmlFor="username">Username:</label>
                                        <input type="text" className="form-control" id="username" onChange={this.handleUsername} />
                                    </div>
                                )}
                                <div className="form-group text-center">
                                    <button type="submit" className="btn btn-primary form-control">{btnText}</button>
                                </div>
                            </form>
                            {signUp ?
                                <div className="container text-center">
                                    <p>Already have an account? <Link to="/signin">Log In</Link></p>
                                </div>
                                :
                                <div className="container text-center">
                                    <p>Do not have an account? <Link to="/signup">Sign Up</Link></p>
                                </div>
                            }
                        </div>
                    </div>
                </div>
            </div>
        );
    }
};

const mapStateToProps = (state) => {
    return {
        isAuthenticated: state.CurrentUser.isAuthenticated,
        error: state.Errors.error
    };
};

export default connect(mapStateToProps, { userAuthentication, addError })(AuthForm);