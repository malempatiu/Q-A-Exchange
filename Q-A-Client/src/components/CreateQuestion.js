import React, { Component } from 'react';
import {connect} from 'react-redux';
import {createQuestion} from '../actions/actions';
import {Redirect} from 'react-router-dom';

export class CreateQuestion extends Component {
    state = {
        text: ''
    };
    handleCreateQuestion = (e) => {
      this.setState({text: e.target.value});
    };
    handleFormSubmit = (e) => {
       e.preventDefault();
       const {text} = this.state;
       this.props.createQuestion({text, createdBy:this.props.email});
       const {error} = this.props;
       if(!error){
         this.props.history.push('/');
       } 
    };
    render() {
        const {error, isAuthenticated} = this.props;
        if(!isAuthenticated){
            return <Redirect to='/signin' />
         };
        return (
            <div className="container addquestion-cont">
                <h4 className="text-center">Ask a question</h4>
                {error && <div className="alert alert-danger">{error}</div>}
                <div className="row">
                    <div className="col-lg-12 col-md-12 col-sm-12">
                        <form onSubmit={this.handleFormSubmit}>
                            <div className="form-group">
                                <textarea className="form-control" id="exampleFormControlTextarea1" rows="3" onChange={this.handleCreateQuestion}></textarea>
                            </div>
                            <div className="form-group text-center">
                                <button type="submit" className="btn btn-primary form-control">Create a Question</button>
                            </div>
                        </form>
                    </div>
                </div>
            </div>
        );
    };
};

const mapStateToProps = (state) => {
    return {
       error: state.Errors.error,
       isAuthenticated: state.CurrentUser.isAuthenticated,
       email: state.CurrentUser.user.email
    }
};

export default connect(mapStateToProps, {createQuestion})(CreateQuestion);
