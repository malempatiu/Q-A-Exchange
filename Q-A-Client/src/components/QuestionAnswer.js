import React, { Component } from 'react';
import { connect } from 'react-redux';
import { getQuestionAndAnswer, createAnswer } from '../actions/actions';
import Moment from 'react-moment';

class QuestionAnswer extends Component {
    state = {
        answer: '',
        createdBy: ''
    };
    componentDidMount() {
        const id = this.props.match.params.id;
        this.props.getQuestionAndAnswer(id);
    };
    handleQuestionAnswer = (e) => {
         this.setState({answer: e.target.value});
    };
    handleUsername = (e) => {
         this.setState({createdBy: e.target.value});
    };
    handleFormSubmit = (e) => {
        e.preventDefault();
        const id = this.props.match.params.id;
        this.props.createAnswer(this.state, id);
        document.getElementById("answerform").reset();
    };
    render() {
        const { data, error } = this.props;
        return (
            <div className="container answers-cont">
                {error && <div className="alert alert-danger">{error}</div>}
                <h3 className="font-weight-bold">{data.text}</h3>
                <p className="text-muted">CreatedBy: {data.createdBy}</p>
                <p className="text-muted">
                    <Moment className="text-muted" format="Do MMM YYYY">
                        {data.createdAt}
                    </Moment>
                </p>
                <hr />
                <h3 className="font-weight-bold">Answers</h3>
                {data.answers.map((qAnswer) => {
                    return <div key={qAnswer._id} className="row">
                        <div className="col-lg-12 col-md-12 col-sm-12">
                            <div className="card">
                                <div className="card-body">
                                    <p className="text-dark">{qAnswer.answer}</p>
                                    <p className="text-muted">Created By: {qAnswer.createdBy}</p>
                                    <Moment className="text-muted" format="Do MMM YYYY">
                                        {qAnswer.createdAt}
                                    </Moment>
                                </div>
                            </div>
                        </div>
                    </div>
                })}
                <hr />
                <h3 className="font-weight-bold post-answer">Your Answer</h3>
                <div className="row">
                    <div className="col-lg-12 col-md-12 col-sm-12">
                        <form id="answerform" onSubmit={this.handleFormSubmit}>
                            <div className="form-group">
                                <textarea className="form-control" id="exampleFormControlTextarea1" rows="3" onChange={this.handleQuestionAnswer}></textarea>
                            </div>
                            <div className="form-group">
                                <input className="form-control" type="text" placeholder="enter email address" onChange={this.handleUsername} /> 
                            </div>
                            <div className="form-group text-center">
                                <button type="submit" className="btn btn-primary form-control">Post Answer</button>
                            </div>
                        </form>
                    </div>
                </div>
            </div>
        );
    };
};

const mapStateToProps = (state) => {
    return {
        data: state.QuestionAndAnswer,
        error: state.Errors.error
    };
};

export default connect(mapStateToProps, { getQuestionAndAnswer, createAnswer })(QuestionAnswer);