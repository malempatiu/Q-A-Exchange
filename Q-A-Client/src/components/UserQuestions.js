import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchUserQuestions, removeError} from '../actions/actions';
import { Link } from 'react-router-dom';

export class UserQuestions extends Component {
    state = {
        text: ''
    }
    componentDidMount() {
        this.props.removeError();
        this.props.fetchUserQuestions();
    };
    handleUserSearch = (e) => {
        this.setState({ text: e.target.value });
    };
    render() {
        const { username, questions, error } = this.props;
        return (
            <div className="conatiner homepage">
                <h1 className="text-center"><span className="font-weight-bold text-dark">{username}</span> your questions</h1>
                <div className="container">
                    {error && <div className="alert alert-danger">{error}</div>}
                    {!questions.length && <div className="alert alert-info">You asked zero questions</div>}
                    {questions.map((question) => {
                        return <div key={question._id} className="row">
                            <div className="col-lg-12 col-md-12 col-sm-12">
                                <div className="card">
                                    <Link to={`/question/${question._id}`} id={question._id} className="card-body text-primary">
                                        {question.text}
                                    </Link>
                                </div>
                            </div>
                        </div>
                    })}
                </div>
            </div>
        );
    }
};

const mapStateToProps = (state) => {
    return {
        questions: state.UserAskedQuestions.userQuestions,
        error: state.Errors.error,
        username: state.CurrentUser.user.username
    };
}

export default connect(mapStateToProps, { fetchUserQuestions, removeError })(UserQuestions);