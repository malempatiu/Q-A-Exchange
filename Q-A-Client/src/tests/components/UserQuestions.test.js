import React from 'react';
import {UserQuestions} from '../../components/UserQuestions';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

describe('USERQUESTIONS COMPONENT', () => {
     it('should render component and set state text', () => {
         const questions = [
             {
               _id: '1235456852',
               text: 'Node.js'
             },
             {
                _id: '6658816594',
                text: 'React'
             }
         ];
         const removeErrorSpy = jest.fn();
         const fetchUserQuestionsSpy = jest.fn();
         const wrapper = shallow(<UserQuestions questions={questions} removeError={removeErrorSpy} fetchUserQuestions={fetchUserQuestionsSpy} username='ugesh'/>);
         expect(toJson(wrapper)).toMatchSnapshot();
     });
});
