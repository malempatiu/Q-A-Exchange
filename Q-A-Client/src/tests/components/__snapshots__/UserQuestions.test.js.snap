// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`USERQUESTIONS COMPONENT should render component and set state text 1`] = `
<div
  className="conatiner homepage"
>
  <h1
    className="text-center"
  >
    <span
      className="font-weight-bold text-dark"
    >
      ugesh
    </span>
     your questions
  </h1>
  <div
    className="container"
  >
    <div
      className="row"
      key="1235456852"
    >
      <div
        className="col-lg-12 col-md-12 col-sm-12"
      >
        <div
          className="card"
        >
          <Link
            className="card-body text-primary"
            id="1235456852"
            replace={false}
            to="/question/1235456852"
          >
            Node.js
          </Link>
        </div>
      </div>
    </div>
    <div
      className="row"
      key="6658816594"
    >
      <div
        className="col-lg-12 col-md-12 col-sm-12"
      >
        <div
          className="card"
        >
          <Link
            className="card-body text-primary"
            id="6658816594"
            replace={false}
            to="/question/6658816594"
          >
            React
          </Link>
        </div>
      </div>
    </div>
  </div>
</div>
`;
